text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,396 @@
+
+Moving to a unified concept of `identity` opens up a path to clean up these method signatures.
+
+### Operations
+
+Operations are a foundational concept for `acid` transactions. Without the ability to describe an operation and a clear mental model of what
+ operations exist and achieve, it is difficult to understand how an action affects state. Granularity and clarity is key.
+
+While `ember-data` does not yet have concepts of operations or transactions, mutations and updates are applied directly, there are many areas
+ we could improve upon by introducing them. As with data, operations upon data should be serializable so that local state can be accurately
+ cached on local clients.
+
+### Nested Saves / API Transactions / Websocket Support
+
+Many applications wish to create or update and save multiple records together. Achieving this in `ember-data` today is unweildy and has many
+ difficult edge cases: one of which is correctly matching data received back from the API to the newly created records already on the client.
+
+A similar edge case occurs when a newly created record is saved for the first time and prior to receiving the request response the same record
+  is recieved via another means (background polling, websocket subscription, etc.). We have no means of matching the record returned by the alternative means to that of the request, leading to a second cache entry being created and an error once the initial request completes.
+
+One solution has been to generate and assign `id`s for records on the client, but this is not always desireable. These scenarios are a major 
+ motivation for `lid` in the `json-api` spec. Users wishing to solve these cases would be able to serialize the `lid` of the `Identifier`
+ for a newly created record and reflect that `lid` back in any payloads send from their API for the session to correctly match the payloads to the record.
+
+### Better Cache Serialization & Improved Infra for Offline Support
+
+In order to enable users to achieve full offline support, or to serialize the store for transport across the wire (for example as an advanced 
+ fastboot rehydration mode) the entire state of the store needs to be serializable. This RFC introduces the foundation for mechanisms through
+ which this can be later achieved.
+
+----
+
+## Detailed design
+
+```typescript
+const IS_IDENTIFIER = Symbol('is-identifier');
+
+// provided for additional debuggability
+const DEBUG_CLIENT_ORIGINATED = Symbol('record-originated-on-client');
+const DEBUG_IDENTIFIER_BUCKET = Symbol('identifier-bucket');
+
+// json-api payload version
+export interface ResourceIdentifierObject {
+  lid?: string;
+  id: string;
+  type: string;
+}
+
+export interface Identifier {
+  [IS_IDENTIFIER]: true;
+  [DEBUG_IDENTIFIER_BUCKET]: string;
+  lid: string;
+}
```

Comment:
What came of the discussion around using only `lid: string` and storing additional information in a `WeakMap`?

Reply:
As it was discussed further with other folks (including @stefanpenner) the realization was had that we do not need referential stability for WeakMap purposes (both because `lid` is stable and because ember-data has need of controlling most of the object lifecycle).

However, continuing to use an object wrapper, especially a stable object wrapper, comes with a multitude of benefits, including:

- **debugging:** enhanced debugging by associating additional information with the identifier in dev builds
- **debugging:** enhanced debugging by users being able to see `type` and `id` at a glance when
- **bug prevention:** enforcement of the use of the identifier generation process (to ensure lookup tables are properly populated)
- **debugging:** ability to tell at a glance that an identifier was properly processed
- **ergonomics:** closer alignment to jsonapi that makes it true that all `RecordIdentifiers` are also `ResourceIdentifierObjects`

Reply:
I added this reasoning into a discussion in `Alternatives`